added ability to pass in validation as an optional paramater

validkeys opened the pull request:
Thanks for creating this lib!

I wanted to be able to run validations on my virtual fields but was unable to do so given the original setup. This PR adds the ability to pass in validation as an optional second parameter to the "field" method. If a validation object has been passed in, the set method calls the ```this.setDataValue``` method in order to trigger validation.

Let me know if you have any feedback :)

validkeys commented:
There doesn't appear to be a travis.yml file in the master branch off of which i based this PR.

defunctzombie commented on index.js:
```diff
@@ -47,12 +47,23 @@ EncryptedField.prototype.vault = function(name) {
 
     return {
-        type: self.Sequelize.VIRTUAL,
+        type: fieldType,
```
Why wouldn't the field be virtual? We don't want the field itself to be saved as it is just a proxy for the decrypted values from the vault. I am concerned that by not having it be virtual, sequelize will try to load the field.

validkeys replied:
It's not that the field wouldn't be virtual. Virtual fields can optionally take a return type (http://docs.sequelizejs.com/en/latest/api/datatypes/#virtual)

So a generic virtual field would be:

```
Sequelize.VIRTUAL
```

But a virtual field that should return a boolean type would be:

```
Sequelize.VIRTUAL(Sequelize.BOOLEAN...
```

So the actual type will always be virtual, I'm just doing a check to see if the user explicitly included a return type. I will clean up my code and rebase off of master to include the last PR acceptance.

defunctzombie commented:
Please fix your indentation. You used 2 spaces where all of the existing indentation is 4.

validkeys commented on index.js:
```diff
@@ -47,12 +47,20 @@ EncryptedField.prototype.vault = function(name) {
-EncryptedField.prototype.field = function(name) {
-    var self = this;
+EncryptedField.prototype.field = function(name, validation) {
+    var self            = this;
```
sorry -- my OCD got the better of me. Will revert.

validkeys commented on index.js:
```diff
@@ -47,12 +47,20 @@ EncryptedField.prototype.vault = function(name) {
         set: function set_encrypted(val) {
+
+            if (hasValidations) {
+              this.setDataValue(name, val);
```
whoops. will fix. thanks!

validkeys commented:
@defunctzombie so I cleaned up the code and rebased off of master. So to recap, the main additions here are that you get all of the same configuration options are you would when creating a regular sequelize field:

```javascript
    private_2: enc_fields.field('private_2', {
      type: Sequelize.TEXT,
      validate: {
        isInt: true  
      },
      allowNull: false,
      defaultValue: null
    })
```

If you pass in the "type", your field will be cast as that type using the built-in functionality of sequelize. You can pass in validation rules just like you would a regular field and you can also supply a default value. One of the issues I was having was when creating models where the field is not in the payload, the encrypted field was showing up like this:

```
{"encrypted": {"someField":"undefined"}}
```

So passing in the the default value, the getter on the virtual field returns your defaultValue if the value in the encrypted field is empty.

Finally, the ability to pass in the allowNull attr as you would a regular property.

I've also added tests for each of the above.

defunctzombie commented:
@validkeys awesome explanation. I understand now and this is a great improvement. Could you please squash the commits down to one right now the diff seems wrong to me (somehow one of the commits from master made it into this PR?)

validkeys commented:
@defunctzombie something happened after I rebased. It included files added from your master commits. My git abilities are somewhat lacking -- not sure what to do at this point. If you've got any ideas, let me know. Otherwise, I'll search around to see if I can find what's happening.

validkeys commented:
@defunctzombie it looks like it might just be github's view here that's showing the files that were changed. The actual pull request only has the 3 changed files.

![screen shot 2016-10-01 at 4 41 05 pm](https://cloud.githubusercontent.com/assets/21793/19016910/f987aa82-87f5-11e6-834c-efefb9377fde.png)

validkeys commented:
@defunctzombie I just keep digging myself deeper. I am going to close this PR and submit a new one. Sorry about this. GIT has always been my weakness :)
